### ISSUE TITLE
Not enough space at the bottom of the page to set Link as active

### ISSUE THREAD

**Issue by ShiiRochi** [fisshy/react-scroll#335]
I have components SideMenu and BlocksRenderer. In SideMenu, Links are rendered and by clicking on them we're scrolling to the relative Element in BlocksRenderer. I'm using 'spy' mode to automatically highlight Links while mouse scrolling. However, I haven't enough space after last two blocks, i.e. page's height is not enough and their links are never highlighted as active. 
What can or should I do to overcome this issue and set them active after clicking on them and at the same time using spy mode on all Links? Maybe there is a way to pass Link instance through the event and set it active after the scroll to that Element is completed?
P.S. Blocks can have different height.
P.P.S. There could be an unlimited amount of blocks.

**Comment by fisshy** [comment]
@thm-design thats usually what I do as well, if the scroll can't scroll down to the element, then it's impossible to determine the active link, unless it's forced by the clicked link, but then it would be interrupted once the scrollbar was moved.

**Comment by ShiiRochi** [comment]
Is there a way to set link's active state in case the user's scroll position is exactly between two elements?

**Comment by fisshy** [comment]
Not sure if its doable, but you could expand 2 and put 3 inside 2.

**Comment by AdamGerthel** [comment]
Perhaps something like [VisSense](https://github.com/vissense/vissense) could be used to determine if one section is more visible than another?

**Comment by ndinhphi** [comment]
I have same isssue

**Comment by online0227** [comment]
still not fixed properly?

**Comment by ckeeney** [comment]
A slightly cleaner solution is to ensure the last section is tall enough.  I gave all my sections a `.section` class and then added to the container:
```css
  .section:last-child {
    min-height: 100vh;
  }
```

**Comment by ShiiRochi** [comment]
I’am not working close with project, that faced that issue anymore, but still this issue is opened.  

I have the following thoughts on the matter.  

So...to determine what item X from N visible items should be highlighted there should be some criteria.  

For example, when 3 items are visible at the same time and all are equally dividing currently visible viewport between each other...there is only one question from the point of view of the package: “which one should I highlight?”  

Also, there could be one more question: “may I highlight all visible?”, but this is another story...  

Currently, I think, package has a solid logic about it.  

We, as users, who see application could say: “If user viewport has reached the bottom of a page, then I want to highlight the most bottom link”. However, how the package could determine when to highlight block above the most bottom one.  

If being at the bottom of the page last item should be highlighted, then it can result in a case, where pixels region, where block above should be highlighted, would be very wise for us, so it will be a pixel contest to catch a right position when block is highlighted (even some .000000003px may be important).  

Thus...the most simple solution is to keep enough distance between blocks, between blocks section start and first block, and between blocks section end and last block.  

Other possible solutions:

1) Library could also accepts some breakpoints, where to highlight each block by default, but it could result in a complex logic.  

2) Also...library can allow user to determine what item should be highlighted by passing info in a callbacks about block visibility.

**Comment by nick4fake** [comment]
What about simply adding an option like "bottomElement" and make it active if scrolled to bottom?

**Comment by marksy** [comment]
yeah i made a dummy component `<LastPage />` with `height:54vh` to fill the bottom void:

```
<Pages>
  {data.map((page: PageDataProps, index: number) => (
    <PageItem key={`page-${page.id}`} isFixedSize={isFixedSize}>
      {page.isLoading ? (
        <Page id={page.id}>
           <StyledLoadingSpinner color={'black'} justify={'space-between'} />
        </Page>
        ) : (
        <Page id={page.id} />
        )} 
    </PageItem>
  ))}
  <LastPage />
</Pages>
```

**Comment by LucianoTrujillo** [comment]
Please this is a really important fix!

**Comment by FrancoMuniz** [comment]
Has anyone found any nice fix for this? Same thing as everyone, I'm adding extra margin at the bottom to toggle last section's Link.

**Comment by iraqwarvet31** [comment]
Having the same issue

**Comment by fisshy** [comment]
I'm not sure how this issue should be fixed, And I welcome any PR solving it.

**Comment by anand-121** [comment]
One solution can be, we set a negative offset value that will make the last element active even it is not at the top. And, disable the navigation on click rather use onClick props and manually navigate to the selected element.
I had also asked a sage question on this - https://stackoverflow.com/questions/74753194/react-customise-react-scroll-behaviour

**Comment by darioackermann** [comment]
Hi all

In case this is still relevant for anyone, create a hook, I called it `useOnScreen`

```ts
import {useEffect, useMemo, useState} from "react";

export default function useOnScreen(ref: HTMLElement) {

    const [isIntersecting, setIntersecting] = useState(false)

    const observer = useMemo(() => new IntersectionObserver(
        ([entry]) => setIntersecting(entry.isIntersecting)
    ), [ref])


    useEffect(() => {
        if(ref)
            observer.observe(ref)
        return () => observer.disconnect()
    }, [])

    return isIntersecting
}
```
Then, at the place where you use your `Link`

```ts
const SidebarScrollItem = ({ title, icon, to }: SidebarScrollLinkItemProps) => {
   // ....
  const isVisible = useOnScreen(document.getElementsByName(to)[0]);
  return (
      <Link
        className={`nav-link ${isVisible ? 'active' : ''}`}
        role="button"
        to={to}
        containerId="main"
        smooth={true}
      >
         // ...
      </Link>
  );
```

**Comment by darioackermann** [comment]
Hi all

In case this is still relevant for anyone, create a hook, I called it useOnScreen

```ts
import {useEffect, useMemo, useState} from "react";

export default function useOnScreen(name: string) {

    const [elem, setElem] = useState<HTMLElement>();
    const [isIntersecting, setIntersecting] = useState(false)

    const mutationObserver = useMemo(() => new MutationObserver(() => {
        const elem = document.getElementsByName(name)[0];
        if(elem)
            setElem(elem)
        }), [name]
    )
    const observer = useMemo(() => new IntersectionObserver(
        ([entry]) => setIntersecting(entry.isIntersecting)
    ), [])


    useEffect(() => {
        if(elem)
            observer.observe(elem)
        else
            mutationObserver.observe(document, {subtree: true, childList: true})
        return () => observer.disconnect()
    }, [elem, observer])

    return isIntersecting
}
```


Then, at the place where you use your Link

```ts
const SidebarScrollItem = ({ title, icon, to }: SidebarScrollLinkItemProps) => {
   // ....
  const isVisible = useOnScreen(to);
  return (
      <Link
        className={`nav-link ${isVisible ? 'active' : ''}`}
        role="button"
        to={to}
        containerId="main"
        smooth={true}
      >
         // ...
      </Link>
  );
```
Hope it helps ::)
